src/os/bluestore/bluefs_types.h

```diff
@@ -59,51 +59,55 @@ struct bluefs_fnode_delta_t {
 
   DENC_HELPERS
 
-  void bound_encode(size_t& p) const {
-    _denc_friend(*this, p);
-  }
-  void encode(ceph::buffer::list::contiguous_appender& p) const {
-    DENC_DUMP_PRE(bluefs_fnode_t);
-    _denc_friend(*this, p);
-  }
-  void decode(ceph::buffer::ptr::const_iterator& p) {
-    DENC_START(2, 2, p);
+  template <typename P>
+  void encode_helper(P& p) const {
+    __u8 need_v = type == WAL_V2 ? 2 : 1;
+    DENC_START(need_v, need_v, p);
     denc_varint(ino, p);
     denc_varint(size, p);
     denc(mtime, p);
     denc(offset, p);
+    if (struct_v >= 2) {
+      denc(type, p);
+    } else {
+      // v1 had a hint field that is no longer used
+      uint8_t __unused__ = 0;
+      denc(__unused__, p);
+    }
     denc(extents, p);
-
-    denc(type, p);
-    if (struct_v >= 2 && type == WAL_V2) {
-      ceph_assert(struct_v == 2 && struct_compat == 2);
+    if (struct_v >= 2) {
       denc(wal_limit, p);
       denc(wal_size, p);
     }
     DENC_FINISH(p);
 
   }
 
-  template<typename T, typename P>
-  friend std::enable_if_t<std::is_same_v<bluefs_fnode_delta_t, std::remove_const_t<T>>>
-  _denc_friend(T& v, P& p) {
-    uint8_t version = 1, compat = 1;
-    if (v.type == WAL_V2) {
-      version = 2;
-      compat = 2;
+  void bound_encode(size_t& p) const {
+    encode_helper(p);
+  }
+  void encode(ceph::buffer::list::contiguous_appender& p) const {
+    encode_helper(p);
+  }
+  void decode(ceph::buffer::ptr::const_iterator& p) {
+    DENC_START(0, 2, p);
```

Comment:
This still misses the throw/assert from DECODE_START

Reply:
I know, I skip this aspect of separate work for ceph that will address lack of compat check.

Reply:
so two things to do extra: transactions as a way to really check version, because no assert will be thrown here, and new pr to add checks to `DENC_...` api

Reply:
Correct. We need to use bluefs_transaction_t for protection against malicious downgrading.